ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,567 @@
+   */
+  public SortedMap<String, Long> getUsedCapacityBySegment() {
+    SortedMap<String, Long> usedCapacityBySegments = new TreeMap<>();
+    LogSegment logSegment = log.getFirstSegment();
+    while (logSegment != null) {
+      usedCapacityBySegments.put(logSegment.getName(), logSegment.getEndOffset());
+      logSegment = log.getNextSegment(logSegment);
+    }
+    return usedCapacityBySegments;
+  }
+
+  /**
+   * Gets the total size of valid data for all log segments
+   * @param timeRange the time range for the expected output. Defines both the reference time and the acceptable resolution.
+   * @return a {@link Pair} whose first element is the exact time at which the stats are valid and whose second
+   * element is the total valid data size.
+   */
+  public Pair<Long, Long> getValidDataSize(TimeRange timeRange) {
+    checkForInitialScan();
+    Long allSegmentValidDataSize = 0L;
+    Long lastBucketEndTime = processTimeRange(timeRange);
+    if (lastBucketEndTime != -1) {
+      synchronized (swapLock) {
+        SortedMap<String, Long> map = scanResults.getValidDataSizePerSegment(lastBucketEndTime);
+        for (Long value : map.values()) {
+          allSegmentValidDataSize += value;
+        }
+      }
+      return new Pair<>(lastBucketEndTime / Time.MsPerSec, allSegmentValidDataSize);
+    } else {
+      return outOfBoundValidDataSizeRequest(timeRange);
+    }
+  }
+
+  /**
+   * Gets the size of valid data at a particular point in time for all log segments. The caller specifies a reference time and
+   * acceptable resolution for the stats in the form of a {@link TimeRange}. The store will return data for a point in time within
+   * the specified range.
+   * @param timeRange the time range for the expected output. Defines both the reference time and the acceptable resolution.
+   * @return a {@link Pair} whose first element is the exact time at which the stats are valid and whose second element is the valid data
+   * size for each segment in the form of a {@link SortedMap} of segment names to valid data sizes.
+   */
+  public Pair<Long, SortedMap<String, Long>> getValidDataSizeBySegment(TimeRange timeRange) {
+    checkForInitialScan();
+    SortedMap<String, Long> map;
+    Long lastBucketEndTime = processTimeRange(timeRange);
+    if (lastBucketEndTime != -1) {
+      synchronized (swapLock) {
+        map = scanResults.getValidDataSizePerSegment(lastBucketEndTime);
+      }
+      return new Pair<>(lastBucketEndTime / Time.MsPerSec, map);
+    } else {
+      return outOfBoundValidDataSizeBySegmentRequest(timeRange);
+    }
+  }
+
+  /**
+   * Gets the valid data size of all containers in this blob store.
+   * @return nested {@link HashMap} of serviceIds to containerIds to their corresponding valid data size
+   */
+  public HashMap<String, HashMap<String, Long>> getValidDataSize() {
+    checkForInitialScan();
+    Long referenceTime = time.milliseconds();
+    if (referenceTime > servingWindowBoundary) {
```

Comment:
Sure, I think I didn't put enough thought into it when I introduced on demand scanning. I do have all the information to check if there is an ongoing scan and whether or not that ongoing scan is going to be sufficient for the request.